src/net.cpp

```diff
@@ -328,42 +329,67 @@ bool IsLocal(const CService& addr)
     return mapLocalHost.count(addr) > 0;
 }
 
-CNode* CConnman::FindNode(const CNetAddr& ip)
+CNode* CConnman::FindNode(const CNetAddr& addr, std::optional<bool> inbound)
 {
-    LOCK(m_nodes_mutex);
+    AssertLockHeld(m_nodes_mutex);
     for (CNode* pnode : m_nodes) {
-      if (static_cast<CNetAddr>(pnode->addr) == ip) {
+        if (static_cast<CNetAddr>(pnode->addr) == addr && (!inbound.has_value() || pnode->IsInboundConn() == inbound.value())) {
             return pnode;
         }
     }
     return nullptr;
 }
 
-CNode* CConnman::FindNode(const std::string& addrName)
+CNode* CConnman::FindNode(const std::string& addr, std::optional<bool> inbound)
 {
-    LOCK(m_nodes_mutex);
+    AssertLockHeld(m_nodes_mutex);
     for (CNode* pnode : m_nodes) {
-        if (pnode->m_addr_name == addrName) {
+        if (pnode->m_addr_name == addr && (!inbound.has_value() || pnode->IsInboundConn() == inbound.value())) {
             return pnode;
         }
     }
     return nullptr;
 }
 
-CNode* CConnman::FindNode(const CService& addr)
+CNode* CConnman::FindNode(const CService& addr, std::optional<bool> inbound)
 {
-    LOCK(m_nodes_mutex);
+    AssertLockHeld(m_nodes_mutex);
     for (CNode* pnode : m_nodes) {
-        if (static_cast<CService>(pnode->addr) == addr) {
+        if (static_cast<CService>(pnode->addr) == addr && (!inbound.has_value() || pnode->IsInboundConn() == inbound.value())) {
             return pnode;
         }
     }
     return nullptr;
 }
 
-bool CConnman::AlreadyConnectedToAddress(const CAddress& addr)
+bool CConnman::AlreadyConnectedToAddress(const CAddress& addr, ConnectionType conn_type, bool log)
 {
-    return FindNode(static_cast<CNetAddr>(addr)) || FindNode(addr.ToStringAddrPort());
+    if (Params().IsTestChain()) return false; // allow bilateral connections on test chains
```

Comment:
```cpp
    if (Params().IsTestChain()) return false; // allow bilateral connections on test chains
```

This seems very out of place to me, and would also prevent writing certain types of tests?

Reply:
Which tests?

Reply:
> They would need to be updated not to make bilateral connections.

I would suggest updating the tests to account for the behaviour you've changed (it may also help to make the bug fixes clearer for reviewers); rather than introducing per-chain, networking functionality. This seems generally dangerous, and mostly unintuitive.